File: plugins/ide-features-trainer/src/org/jetbrains/plugins/feature/suggester/defaultSuggesters/SafeDeleteSuggester.scala
package org.jetbrains.plugins.feature.suggester.defaultSuggesters

import org.jetbrains.plugins.feature.suggester.{NoSuggestion, Suggestion, FeatureSuggester}
import org.jetbrains.plugins.feature.suggester.changes.{UserAnAction, ChildRemovedAction, UserAction}
import com.intellij.psi.{PsiField, PsiClass, PsiMethod}
import com.intellij.openapi.command.CommandProcessor
import com.intellij.ide.{IdeTooltipManager, ClipboardSynchronizer}
import java.awt.datatransfer.DataFlavor

/**
 * @author Alefas
 * @since 24.05.13
 */
class SafeDeleteSuggester extends FeatureSuggester {
  val POPUP_MESSAGE = "Why not to try Safe Delete action (Alt + Delete)"

  private var lastTimeForPopup = 0L

  def getSuggestion(actions: List[UserAction], anActions: List[UserAnAction]): Suggestion = {
    val name = CommandProcessor.getInstance().getCurrentCommandName
    if (name != null) return NoSuggestion //it's not user typing action, so let's do nothing

    actions.last match {
      case ChildRemovedAction(parent, child) =>
        child match {
          case _: PsiMethod | _: PsiClass | _: PsiField =>
            val contents = ClipboardSynchronizer.getInstance().getContents
            if (contents != null) {
              val delta = System.currentTimeMillis() - lastTimeForPopup
              if (delta < 50) {
                IdeTooltipManager.getInstance().hideCurrentNow(false)
                return NoSuggestion //do not suggest in case of deletion for few things
              }
              try {
                val clipboardContent = contents.getTransferData(DataFlavor.stringFlavor).asInstanceOf[String]
                if (clipboardContent.contains(child.getText)) return NoSuggestion //this is action with copy to clipboard side effect, so we shouldn't suggest anything
              }
              catch {
                case ignore: Exception =>
              }
              lastTimeForPopup = System.currentTimeMillis()
              return SuggestingUtil.createSuggestion(None, POPUP_MESSAGE)
            }
          case _ =>
        }
      case _ =>
    }
    NoSuggestion
  }

  def getId: String = "Safe delete suggester"
}

File: plugins/ide-features-trainer/src/main/kotlin/org/jetbrains/plugins/feature/suggester/suggesters/SafeDeleteSuggester.kt
package org.jetbrains.plugins.feature.suggester.suggesters

import com.intellij.ide.IdeTooltipManager
import com.intellij.openapi.command.CommandProcessor
import com.intellij.psi.PsiClass
import com.intellij.psi.PsiDeclarationStatement
import com.intellij.psi.PsiField
import com.intellij.psi.PsiMethod
import org.jetbrains.plugins.feature.suggester.FeatureSuggester
import org.jetbrains.plugins.feature.suggester.NoSuggestion
import org.jetbrains.plugins.feature.suggester.Suggestion
import org.jetbrains.plugins.feature.suggester.history.UserActionsHistory
import org.jetbrains.plugins.feature.suggester.history.UserAnActionsHistory
import org.jetbrains.plugins.feature.suggester.changes.ChildRemovedAction

class SafeDeleteSuggester : FeatureSuggester {
    companion object {
        const val POPUP_MESSAGE = "Just to be on the safe side, try Safe Delete instead (Alt + Delete)"
        const val suggestionCountdownMillis = 150
    }

    private var lastTimeForPopupMillis = 0L

    /**
     * Returns PopupSuggestion when user removing variable/field/method/class declaration
     * (CHANGED FROM LAST VERSION)
     */
    override fun getSuggestion(actions: UserActionsHistory, anActions: UserAnActionsHistory): Suggestion {
        val name = CommandProcessor.getInstance().currentCommandName
        if (name != null) {
            //it's not user typing action, so let's do nothing
            return NoSuggestion
        }
        val lastAction = actions.last()
        if (lastAction is ChildRemovedAction && isDeclarationRemoving(lastAction)) {
            val delta = System.currentTimeMillis() - lastTimeForPopupMillis
            if (delta < suggestionCountdownMillis) {
                IdeTooltipManager.getInstance().hideCurrentNow(false)
                return NoSuggestion //do not suggest in case of deletion for few things
            }
            lastTimeForPopupMillis = System.currentTimeMillis()
            return createSuggestion(null, POPUP_MESSAGE)
        }
        return NoSuggestion
    }

    private fun isDeclarationRemoving(action: ChildRemovedAction): Boolean {
        val (parent, child) = action
        return parent != null && (child is PsiDeclarationStatement
                || child is PsiField
                || child is PsiMethod
                || child is PsiClass)
    }

    override fun getId(): String = "Safe delete suggester"
}